File: macos-app/Sources/ThoughtInput/Core/OAuthTokenManager.swift
import Foundation

@MainActor
final class OAuthTokenManager {
    static let shared = OAuthTokenManager()

    private var tokenCache: [UUID: OAuthToken] = [:]

    func validToken(for destination: Destination) async throws -> String {
        if let cached = tokenCache[destination.id], !cached.isExpired {
            return cached.accessToken
        }

        // Try refresh if we have a refresh token
        if let cached = tokenCache[destination.id], let existingRefreshToken = cached.refreshToken {
            if let refreshed = try? await refreshToken(existingRefreshToken, for: destination) {
                tokenCache[destination.id] = refreshed
                return refreshed.accessToken
            }
        }

        let token = try await fetchToken(for: destination)
        tokenCache[destination.id] = token
        return token.accessToken
    }

    func clearToken(for destinationID: UUID) {
        tokenCache.removeValue(forKey: destinationID)
    }

    // MARK: - Token Exchange

    private func fetchToken(for destination: Destination) async throws -> OAuthToken {
        switch destination.type {
        case .restOAuthPassword(let config):
            return try await passwordGrant(config: config)
        case .restOAuthClientCredentials(let config):
            return try await clientCredentialsGrant(config: config)
        default:
            CaptureLog.auth.error("fetchToken called for non-OAuth destination type")
            throw OAuthError.missingCredentials
        }
    }

    private func passwordGrant(config: RESTOAuthPasswordConfig) async throws -> OAuthToken {
        guard let username = try KeychainService.loadString(ref: config.usernameRef),
              let password = try KeychainService.loadString(ref: config.passwordRef) else {
            throw OAuthError.missingCredentials
        }

        let body = "grant_type=password&username=\(urlEncode(username))&password=\(urlEncode(password))"
        return try await exchangeToken(tokenURL: config.tokenURL, body: body)
    }

    private func clientCredentialsGrant(config: RESTOAuthClientCredentialsConfig) async throws -> OAuthToken {
        guard let clientID = try KeychainService.loadString(ref: config.clientIDRef),
              let clientSecret = try KeychainService.loadString(ref: config.clientSecretRef) else {
            throw OAuthError.missingCredentials
        }

        let body = "grant_type=client_credentials&client_id=\(urlEncode(clientID))&client_secret=\(urlEncode(clientSecret))"
        return try await exchangeToken(tokenURL: config.tokenURL, body: body)
    }

    private func refreshToken(_ token: String, for destination: Destination) async throws -> OAuthToken {
        let tokenURL: String
        switch destination.type {
        case .restOAuthPassword(let c): tokenURL = c.tokenURL
        case .restOAuthClientCredentials(let c): tokenURL = c.tokenURL
        default: throw OAuthError.missingCredentials
        }

        let body = "grant_type=refresh_token&refresh_token=\(urlEncode(token))"
        return try await exchangeToken(tokenURL: tokenURL, body: body)
    }

    private func exchangeToken(tokenURL: String, body: String) async throws -> OAuthToken {
        guard let url = URL(string: tokenURL) else {
            throw OAuthError.invalidTokenURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        request.httpBody = body.data(using: .utf8)
        request.timeoutInterval = 15

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            let code = (response as? HTTPURLResponse)?.statusCode ?? -1
            throw OAuthError.tokenExchangeFailed(statusCode: code)
        }

        let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        guard let accessToken = json?["access_token"] as? String else {
            throw OAuthError.invalidTokenResponse
        }

        let expiresIn = json?["expires_in"] as? TimeInterval
        let refreshToken = json?["refresh_token"] as? String

        return OAuthToken(
            accessToken: accessToken,
            expiresAt: expiresIn.map { Date.now.addingTimeInterval($0) },
            refreshToken: refreshToken
        )
    }

    private static let formURLEncodedAllowed: CharacterSet = {
        var chars = CharacterSet.alphanumerics
        chars.insert(charactersIn: "-._~")
        return chars
    }()

    private func urlEncode(_ string: String) -> String {
        string.addingPercentEncoding(withAllowedCharacters: Self.formURLEncodedAllowed) ?? string
    }
}

enum OAuthError: LocalizedError, Equatable {
    case missingCredentials
    case invalidTokenURL
    case tokenExchangeFailed(statusCode: Int)
    case invalidTokenResponse

    var errorDescription: String? {
        switch self {
        case .missingCredentials: "OAuth credentials not found in Keychain"
        case .invalidTokenURL: "Invalid token URL"
        case .tokenExchangeFailed(let code): "Token exchange failed (status \(code))"
        case .invalidTokenResponse: "Token response missing access_token"
        }
    }
}

File: android-app/app/src/main/java/com/thoughtinput/capture/data/destinations/OAuthTokenManager.kt
package com.thoughtinput.capture.data.destinations

import com.thoughtinput.capture.data.ApiClient
import com.thoughtinput.capture.util.CaptureLog
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import org.json.JSONObject
import java.net.URLEncoder
import java.util.UUID

sealed class OAuthError(message: String) : Exception(message) {
    object MissingCredentials : OAuthError("OAuth credentials not found in keystore")
    object InvalidTokenURL : OAuthError("Invalid token URL")
    class TokenExchangeFailed(val statusCode: Int) : OAuthError("Token exchange failed (status $statusCode)")
    object InvalidTokenResponse : OAuthError("Token response missing access_token")
}

class OAuthTokenManager(
    private val keystore: SecretsKeystore,
    private val apiClient: ApiClient = ApiClient()
) {
    private val tokenCache = mutableMapOf<UUID, OAuthToken>()
    private val mutex = Mutex()

    suspend fun validToken(destination: Destination): String = mutex.withLock {
        tokenCache[destination.id]?.let { cached ->
            if (!cached.isExpired) return@withLock cached.accessToken
            cached.refreshToken?.let { rt ->
                runCatching { refreshToken(rt, destination) }.getOrNull()?.let { refreshed ->
                    tokenCache[destination.id] = refreshed
                    return@withLock refreshed.accessToken
                }
            }
        }

        val fresh = fetchToken(destination)
        tokenCache[destination.id] = fresh
        fresh.accessToken
    }

    suspend fun clearToken(destinationID: UUID) = mutex.withLock {
        tokenCache.remove(destinationID)
        Unit
    }

    private suspend fun fetchToken(destination: Destination): OAuthToken {
        return when (val type = destination.type) {
            is DestinationType.RestOAuthPassword -> passwordGrant(type)
            is DestinationType.RestOAuthClientCredentials -> clientCredentialsGrant(type)
            else -> {
                CaptureLog.error("Auth", "fetchToken called for non-OAuth destination type")
                throw OAuthError.MissingCredentials
            }
        }
    }

    private suspend fun passwordGrant(config: DestinationType.RestOAuthPassword): OAuthToken {
        val username = keystore.loadString(config.usernameRef)
            ?: throw OAuthError.MissingCredentials
        val password = keystore.loadString(config.passwordRef)
            ?: throw OAuthError.MissingCredentials

        val body = "grant_type=password" +
            "&username=${urlEncode(username)}" +
            "&password=${urlEncode(password)}"
        return exchangeToken(config.tokenURL, body)
    }

    private suspend fun clientCredentialsGrant(config: DestinationType.RestOAuthClientCredentials): OAuthToken {
        val clientID = keystore.loadString(config.clientIDRef)
            ?: throw OAuthError.MissingCredentials
        val clientSecret = keystore.loadString(config.clientSecretRef)
            ?: throw OAuthError.MissingCredentials

        val body = "grant_type=client_credentials" +
            "&client_id=${urlEncode(clientID)}" +
            "&client_secret=${urlEncode(clientSecret)}"
        return exchangeToken(config.tokenURL, body)
    }

    private suspend fun refreshToken(token: String, destination: Destination): OAuthToken {
        val tokenURL = when (val t = destination.type) {
            is DestinationType.RestOAuthPassword -> t.tokenURL
            is DestinationType.RestOAuthClientCredentials -> t.tokenURL
            else -> throw OAuthError.MissingCredentials
        }
        val body = "grant_type=refresh_token&refresh_token=${urlEncode(token)}"
        return exchangeToken(tokenURL, body)
    }

    private suspend fun exchangeToken(tokenURL: String, body: String): OAuthToken {
        if (tokenURL.isBlank()) throw OAuthError.InvalidTokenURL

        val result = apiClient.post(
            url = tokenURL,
            headers = emptyMap(),
            body = body.toByteArray(Charsets.UTF_8),
            contentType = "application/x-www-form-urlencoded",
            timeoutMs = 15_000
        )

        return when (result) {
            is ApiClient.Result.Success -> parseTokenResponse(result.body)
            is ApiClient.Result.HttpError -> throw OAuthError.TokenExchangeFailed(result.statusCode)
            is ApiClient.Result.IoError -> throw OAuthError.TokenExchangeFailed(-1)
        }
    }

    private fun parseTokenResponse(body: String): OAuthToken {
        val json = try { JSONObject(body) } catch (e: Exception) {
            throw OAuthError.InvalidTokenResponse
        }
        val accessToken = json.optString("access_token", "").ifBlank {
            throw OAuthError.InvalidTokenResponse
        }
        val expiresIn = if (json.has("expires_in")) json.optLong("expires_in", -1L).takeIf { it > 0 } else null
        val refreshToken = if (json.has("refresh_token")) json.optString("refresh_token", "").ifBlank { null } else null
        return OAuthToken(
            accessToken = accessToken,
            expiresAt = expiresIn?.let { System.currentTimeMillis() + it * 1000L },
            refreshToken = refreshToken
        )
    }

    private fun urlEncode(s: String): String = URLEncoder.encode(s, "UTF-8")
}
